File: android/app/src/test/java/dev/siliconoptimizer/buddy/LoopbackServer.kt
package dev.siliconoptimizer.buddy

import java.io.ByteArrayOutputStream
import java.net.InetAddress
import java.net.ServerSocket
import java.net.Socket
import java.util.concurrent.CopyOnWriteArrayList
import kotlin.concurrent.thread

/**
 * A real HTTP server on 127.0.0.1, for tests that are about what goes over a socket.
 *
 * The transport is the part of this app that talks to something else, and until now
 * nothing exercised it: reverting the SSE blank-line fix left every test green. So this
 * binds a port, speaks the little of HTTP the Mac speaks, answers whatever the test
 * sets, and records what arrived — including how long between connections, which is how
 * a reconnect policy gets checked rather than assumed.
 */
class LoopbackServer : AutoCloseable {

    data class Reply(
        val status: Int,
        val body: ByteArray,
        val contentType: String = "application/json",
        /** Sent in pieces, the way an event stream arrives. */
        val chunks: List<ByteArray>? = null,
        /** Close the connection straight after answering. */
        val closeAfter: Boolean = true,
    )

    data class Received(
        val method: String,
        val path: String,
        val headers: Map<String, String>,
        val body: String,
        val atMillis: Long,
    )

    private val socket = ServerSocket(0, 50, InetAddress.getByName("127.0.0.1"))
    private val repliesByPath = mutableMapOf<String, Reply>()
    private val received = CopyOnWriteArrayList<Received>()
    private val lock = Object()
    @Volatile private var running = true

    val port: Int get() = socket.localPort

    init {
        thread(isDaemon = true, name = "loopback-server") {
            while (running) {
                val connection = try {
                    socket.accept()
                } catch (error: Exception) {
                    break
                }
                thread(isDaemon = true) { serve(connection) }
            }
        }
    }

    // MARK: - What it answers

    fun reply(path: String, status: Int, body: String) {
        synchronized(lock) {
            repliesByPath[path] = Reply(status, body.toByteArray())
        }
    }

    /** An event stream, optionally delivered in pieces to split lines across reads. */
    fun events(path: String, body: String, splitAt: Int? = null) {
        val bytes = body.toByteArray()
        val chunks = splitAt?.let {
            listOf(bytes.copyOfRange(0, it), bytes.copyOfRange(it, bytes.size))
        }
        synchronized(lock) {
            repliesByPath[path] = Reply(200, bytes, "text/event-stream", chunks)
        }
    }

    /** An event stream sent as raw bytes, for CRLF and multi-byte cases. */
    fun eventBytes(path: String, chunks: List<ByteArray>) {
        synchronized(lock) {
            repliesByPath[path] = Reply(
                200, chunks.fold(ByteArray(0)) { all, part -> all + part },
                "text/event-stream", chunks,
            )
        }
    }

    // MARK: - What it saw

    val requests: List<Received> get() = received.toList()

    fun request(path: String): Received? = received.lastOrNull { it.path == path }

    fun requestCount(path: String): Int = received.count { it.path == path }

    /** The gaps between consecutive requests to a path, in milliseconds. */
    fun gaps(path: String): List<Long> {
        val times = received.filter { it.path == path }.map { it.atMillis }
        return times.zipWithNext { earlier, later -> later - earlier }
    }

    /** Waits until a path has been asked for this many times, or gives up. */
    fun awaitRequests(path: String, count: Int, timeoutMs: Long = 15_000): Boolean {
        val deadline = System.currentTimeMillis() + timeoutMs
        while (System.currentTimeMillis() < deadline) {
            if (requestCount(path) >= count) return true
            Thread.sleep(25)
        }
        return false
    }

    override fun close() {
        running = false
        runCatching { socket.close() }
    }

    // MARK: - The little of HTTP the Mac speaks

    private fun serve(connection: Socket) {
        connection.use { open ->
            val input = open.getInputStream()
            val header = ByteArrayOutputStream()
            var previous = 0
            var blankLines = 0
            while (blankLines < 1) {
                val byte = input.read()
                if (byte < 0) return
                header.write(byte)
                if (byte == '\n'.code && previous == '\r'.code && header.size() >= 4) {
                    val text = header.toString("UTF-8")
                    if (text.endsWith("\r\n\r\n")) blankLines++
                }
                previous = byte
            }
            val headerText = header.toString("UTF-8")
            val lines = headerText.split("\r\n").filter { it.isNotEmpty() }
            val requestLine = lines.first().split(" ")
            val headers = lines.drop(1).mapNotNull { line ->
                val colon = line.indexOf(':')
                if (colon <= 0) null
                else line.substring(0, colon).lowercase() to line.substring(colon + 1).trim()
            }.toMap()

            val length = headers["content-length"]?.toIntOrNull() ?: 0
            val body = ByteArray(length)
            var read = 0
            while (read < length) {
                val count = input.read(body, read, length - read)
                if (count < 0) break
                read += count
            }

            val target = requestLine.getOrElse(1) { "/" }
            val path = target.substringBefore('?')
            received.add(
                Received(
                    method = requestLine.firstOrNull() ?: "GET",
                    path = path,
                    headers = headers,
                    body = String(body, Charsets.UTF_8),
                    atMillis = System.currentTimeMillis(),
                ),
            )

            val reply = synchronized(lock) {
                repliesByPath[path] ?: Reply(404, """{"error":"Unknown endpoint"}""".toByteArray())
            }
            val output = open.getOutputStream()
            val head = buildString {
                append("HTTP/1.1 ${reply.status} ${if (reply.status == 200) "OK" else "Error"}\r\n")
                append("Content-Type: ${reply.contentType}\r\n")
                append("Content-Length: ${reply.body.size}\r\n")
                append("Connection: close\r\n\r\n")
            }
            output.write(head.toByteArray())
            output.flush()
            val pieces = reply.chunks ?: listOf(reply.body)
            for ((index, piece) in pieces.withIndex()) {
                output.write(piece)
                output.flush()
                if (index < pieces.lastIndex) Thread.sleep(30)
            }
        }
    }
}

File: ios/SiliconBuddyTests/LoopbackServer.swift
import Foundation
import Network

/// A real HTTP server on 127.0.0.1, for tests that are about what goes over a socket.
///
/// `URLProtocol` is the usual shortcut, but `URLSession.bytes(for:)` — which is how the
/// app reads an event stream — does not deliver a custom protocol's body. So this binds
/// a port, speaks the little of HTTP the Mac speaks, and answers whatever the test set.
/// It also records what arrived, which is how "what did the client actually send?" gets
/// an answer rather than an assumption.
final class LoopbackServer: @unchecked Sendable {

    struct Reply {
        var status: Int
        var body: String
        var contentType = "application/json"
        /// Sent in pieces with a pause between them, the way an event stream arrives.
        var chunked = false
    }

    struct Received {
        var method: String
        var path: String
        var headers: [String: String]
        var body: Data
    }

    private let listener: NWListener
    private let queue = DispatchQueue(label: "dev.siliconoptimizer.buddy.tests.server")
    private let lock = NSLock()
    private var repliesByPath: [String: Reply] = [:]
    private var receivedRequests: [Received] = []

    private(set) var port: Int = 0

    init() throws {
        let parameters = NWParameters.tcp
        parameters.requiredInterfaceType = .loopback
        parameters.allowLocalEndpointReuse = true
        listener = try NWListener(using: parameters, on: .any)

        let ready = DispatchSemaphore(value: 0)
        listener.stateUpdateHandler = { [weak self] state in
            guard case .ready = state, let self else { return }
            self.port = Int(self.listener.port?.rawValue ?? 0)
            ready.signal()
        }
        listener.newConnectionHandler = { [weak self] connection in
            self?.serve(connection)
        }
        listener.start(queue: queue)
        _ = ready.wait(timeout: .now() + 5)
    }

    deinit { listener.cancel() }

    func stop() { listener.cancel() }

    // MARK: - What it answers

    func reply(_ path: String, _ status: Int, _ body: String) {
        lock.lock()
        repliesByPath[path] = Reply(status: status, body: body)
        lock.unlock()
    }

    func events(_ path: String, _ body: String, chunked: Bool = true) {
        lock.lock()
        repliesByPath[path] = Reply(
            status: 200, body: body, contentType: "text/event-stream", chunked: chunked
        )
        lock.unlock()
    }

    // MARK: - What it saw

    var requests: [Received] {
        lock.lock()
        defer { lock.unlock() }
        return receivedRequests
    }

    func request(to path: String) -> Received? {
        requests.last { $0.path == path }
    }

    // MARK: - The little of HTTP the Mac speaks

    private func serve(_ connection: NWConnection) {
        connection.start(queue: queue)
        read(connection, buffer: Data())
    }

    private func read(_ connection: NWConnection, buffer: Data) {
        connection.receive(minimumIncompleteLength: 1, maximumLength: 65_536) {
            [weak self] data, _, isComplete, error in
            guard let self else { return }
            var buffer = buffer
            if let data { buffer.append(data) }
            if error != nil { return connection.cancel() }

            guard let headerEnd = buffer.range(of: Data("\r\n\r\n".utf8)) else {
                if isComplete { return connection.cancel() }
                return self.read(connection, buffer: buffer)
            }
            let headerText = String(decoding: buffer[..<headerEnd.lowerBound], as: UTF8.self)
            var lines = headerText.components(separatedBy: "\r\n")
            let requestLine = lines.removeFirst().split(separator: " ")
            var headers: [String: String] = [:]
            for line in lines {
                guard let colon = line.firstIndex(of: ":") else { continue }
                headers[line[..<colon].lowercased()] =
                    line[line.index(after: colon)...].trimmingCharacters(in: .whitespaces)
            }
            var body = Data(buffer[headerEnd.upperBound...])
            if let length = headers["content-length"].flatMap(Int.init), body.count < length {
                if isComplete { return connection.cancel() }
                return self.read(connection, buffer: buffer)
            }

            let target = requestLine.count > 1 ? String(requestLine[1]) : "/"
            let path = URLComponents(string: "http://localhost\(target)")?.path ?? target
            self.lock.lock()
            self.receivedRequests.append(
                Received(
                    method: requestLine.first.map(String.init) ?? "GET",
                    path: path, headers: headers, body: body
                )
            )
            let reply = self.repliesByPath[path]
                ?? Reply(status: 404, body: #"{"error":"Unknown endpoint"}"#)
            self.lock.unlock()
            body = Data()
            self.write(reply, to: connection)
        }
    }

    private func write(_ reply: Reply, to connection: NWConnection) {
        let payload = Data(reply.body.utf8)
        let head = [
            "HTTP/1.1 \(reply.status) \(reply.status == 200 ? "OK" : "Error")",
            "Content-Type: \(reply.contentType)",
            "Content-Length: \(payload.count)",
            "Cache-Control: no-store",
            "Connection: close",
        ].joined(separator: "\r\n") + "\r\n\r\n"

        connection.send(content: Data(head.utf8), completion: .contentProcessed { _ in
            guard reply.chunked, payload.count > 32 else {
                connection.send(content: payload, completion: .contentProcessed { _ in
                    connection.cancel()
                })
                return
            }
            // An event stream arrives in pieces, and the parser has to survive a split
            // that lands in the middle of a line.
            let half = payload.count / 2
            connection.send(content: payload.prefix(half), completion: .contentProcessed { _ in
                self.queue.asyncAfter(deadline: .now() + 0.05) {
                    connection.send(
                        content: payload.suffix(from: half),
                        completion: .contentProcessed { _ in connection.cancel() }
                    )
                }
            })
        })
    }
}
